Bluetooth: Mesh: Make network msg cache netkey aware

omkar3141 opened the pull request:
Improve the network message cache to be aware of network keys to prevent false duplicate detection across different subnets. This ensures that messages with the same source address and sequence number but from different network keys are not incorrectly identified as duplicates, as it can happen in certain cases. See ES-26350.

omkar3141 committed:
Bluetooth: Mesh: Make net msg cache netkey aware

Improve the network message cache to be aware of network keys to
prevent false duplicate detection across different subnets. This
ensures that messages with the same source address and sequence
number but from different network keys are not incorrectly
identified as duplicates, as it can happen in certain cases.
See ES-26350.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -82,7 +82,8 @@ struct iv_val {
 	uint32_t src : 15, /* MSb of source is always 0 */
-	      seq : 17;
+		 seq : 17;
+	const struct bt_mesh_net_cred *cred;
```
> Using array pointer can also let a message pass the message cache if secondary key was used to decrypt message M for the first time, but the KR procedure completed, key pointers were [swapped](https://github.com/zephyrproject-rtos/zephyr/blob/2d72d86ba2d561d76243879dd0b29edbf9ab7608/subsys/bluetooth/mesh/subnet.c#L233-L237) and now the same message M relayed with some delay will go in again because `cred` pointers won't match.

I guess after swapping keys, deobfuscation gives garbage address and sequence number since privacy is swapped too.

alxelax replied:
I guess this is not issues of pointer usage. Network cache should be purged after removing subnetwork (friendship termination???). It prevents considering deprecated data for networking.

PavelVPV replied:
> Network cache should be purged after removing subnetwork (friendship termination???).

But this is not the case for key refresh.

Purging or updating message cache may work, but I guess there is some extra risk in this as well.

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -653,15 +657,14 @@ static bool net_decrypt(struct bt_mesh_net_rx *rx, struct net_buf_simple *in,
 	}
 
-	if (rx->net_if == BT_MESH_NET_IF_ADV && msg_cache_match(out)) {
+	if (rx->net_if == BT_MESH_NET_IF_ADV && msg_cache_match(out, rx)) {
```
I guess it is better to pass `net_idx` directly. `msg_cache_match` doesn't need to know about `bt_mesh_net_rx`.

PavelVPV replied:
This is not addressed

omkar3141 replied:
Haah ... changes were not staged :|

jhedberg commented:
This doesn't look like a bug fix, so most likely Zephyr 4.4 material rather than 4.3. Do you agree? If so, could you set the 4.4 milestone, so that this doesn't show up in the merge-list until after the final release?

omkar3141 committed:
Bluetooth: Mesh: Test network cache improvement

Adds a test that checks that the network message cache differentiates
between PDUs with the same SRC and SEQ but different NetKey Index.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@nordicsemi.no>

omkar3141 commented on tests/bluetooth/mesh/net/CMakeLists.txt:
This is unfortunately needed because net.c has for some unknown reasons [this](https://github.com/omkar3141/zephyr/blob/b646c1b7b49bc8fc1b275a8519bdba8166576b59/subsys/bluetooth/mesh/net.c#L93-L94) and thus requires "[sar_cfg_internal.h](https://github.com/omkar3141/zephyr/blob/b646c1b7b49bc8fc1b275a8519bdba8166576b59/subsys/bluetooth/mesh/net.c#L38)" to be included.

I think we can clean this part of `net.c` separately in another PR.

PavelVPV commented on tests/bluetooth/mesh/net/src/main.c:
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2025 Nordic Semiconductor ASA
```
